rfcs/text/0000-rfc-library-preparation.md

```diff
@@ -0,0 +1,428 @@
+
+---
+
+### Implementation changes for DCP components
+
+#### Ingestion Service
+
+- Logical units will be anchored on a library preparation entity UUID (or possibly the process entity that is input to the library preparation entity) in the experimental graph (exporter).
+- Schema template generator will be modified to (1) put the library preparation entity fields in the Sequence file tab, and (2) exclude all the library preparation entity fields except the required ones. 
+- Experimental graph (instantiated by links.json) will represent that (1) a library preparation protocol is followed to produce a library preparation from a cell suspension (or other biomaterial for bulk sequencing) and that (2) a sequencing protocol is followed to produce sequence file(s) from a library preparation. Users of the ingest-api will supply these links during programmatic submission. Otherwise, Ingestion Service (ingest-broker) will create the following links from metadata in the Sequence file tab and some graph assumptions (Fig. 6):
+    - Cell suspension :left_right_arrow: Process_1 :left_right_arrow: Library preparation :left_right_arrow: Process_0 :left_right_arrow: Sequence file
+    - Process_1 :left_right_arrow: Library preparation protocol
+    - Process_0 :left_right_arrow: Sequencing protocol
+    - Declare rules about linking
+    
+---
+    
+![Figure 6](/rfcs/images/0000-lib_prep_rfc_fig6.png)
+
+**Figure 6: Diagram of how protocol linking will change with new library preparation entity**. A) Current metadata model showing that links to Sequencing protocol and Library preparation protocol entities are generated from the last process (arrow) in the graph. B) New metadata model showing that a link to the Sequencing protocol is generated from the ultimate process (process_0) in the graph, and a link to the Library preparation protocol is generated from the penultimate process (process_1) in the graph.
+
+---
+
+#### Data Storage Service
+
+- No changes will be needed.
+- Clients may need to update queries and subscriptions (some described below).
+
+#### Upload Service
+
+- No changes will be needed.
+
+#### Data Processing Pipelines
+
+- Subscription queries will reference a library preparation entity instead of cell suspension entity, where appropriate.
+- If multiple-bundles:1-workflow remains a use case, pipelines can query for all bundles with a particular library preparation UUID in order to trigger a single workflow.
+
+#### Data Brower
+
+- Is anything coded against cell suspension?
+- How will this affect displaying rows in Sample view?
+
+#### Matrix Service
+
+- Is anything coded against cell suspension?
```

Comment:
> The matrix service (and azul/data browser I believe) does make the assumption that the "lowest" biomaterial for a given cell is a cell_suspension. So there will be a little work required to shift that down to library prep. 

I appreciate this change might need to be made. I don't know exactly what the assumptions are that these services make, but I wonder if cell suspension could still be used for some things (like assigning cell IDs) or whether it just makes it easier to shift everything down to LP.

Reply:
> Also, it might get a little confusing if we use "library prep" as a shorthand to refer to both the library preparation protocol and the library preparation biomaterial.

I agree that it's confusing to have both "Library preparation protocol" and "Library preparation" (the biomaterial). Based on my background of both working with and making library preparations, I feel these are the most "accurate" terms to describe these concepts, but am open to doing some research to see if there are alternative, but still accurate, terms to use. This RFC could probably be approved w/o resolving the naming issue.

Reply:
I want to actually do some research to see how submitters/researchers refer to this entity.
I also want to see how they keep and differentiate this information themselves.

Am I right that the single cell expression atlas call this thing an assay? (not saying that's the right name, just wondered if it's the same thing.)

If there are any current submitters who will benefit from having the library prep ID to correctly represent their experiment (if they did 2 libraries, e.g.), maybe we can briefly chat to them?

Lastly, what is the assumption - do they keep this ID themselves and we're benefiting from this, or are we making them make it up on the spot?

Gabs.

Reply:
> Am I right that the single cell expression atlas call this thing an assay?

Based on the fact that in 1 project I checked for the SCEA, they use the term "Assay" to refer to 1 "run". A library preparation (biomaterial) can have 1 or more runs.

> If there are any current submitters who will benefit from having the library prep ID to correctly represent their experiment, maybe we can briefly chat to them?

Don't see why not! Ask in the wrangling channel, as I'm not working with any current contributors.

> what is the assumption - do they keep this ID themselves and we're benefiting from this, or are we making them make it up on the spot?

For the most part, I imaging if a simple 1 cell suspension -> 1 library prep experiment is done, then the ID they keep track of might be the same for both of these. If the design is 1 cell suspension -> 1+ library preps, I imagine they would have a new ID for the LPs (maybe by adding "rep1", "rep2" or something like that to the end of the cell suspension ID).